src/proc/unused.rs

```diff
@@ -0,0 +1,386 @@
+    for entry_point in entry_points {
+        let ctx = &mut CrawlContext {
+            module,
+            function: &entry_point.function,
+            used_global_variables: &mut used_global_variables,
+            used_types: &mut used_types,
+            used_constants: &mut used_constants,
+            used_functions: &mut used_functions,
+        };
+
+        // Add all argument types to the used types list
+        for arg in &entry_point.function.arguments {
+            crawl_type(arg.ty, &ctx.module.types, ctx.used_types);
+        }
+
+        // Crawl the function body and add any global variables and typesthat are used to the used list
+        crawl_block(&entry_point.function.body, ctx);
+    }
+
+    // For every used global variable
+    for global_variable in &used_global_variables {
+        // Add it's type to the list of used types
+        let var = &module.global_variables[*global_variable];
+        crawl_type(var.ty, &module.types, &mut used_types);
+    }
+
+    // Any items that are not in the used list, add to the unused list
+    for (handle, _) in module.types.iter() {
+        if !used_types.contains(&handle) {
+            unused.types.insert(handle);
+        }
+    }
+    for (handle, _) in module.global_variables.iter() {
+        if !used_global_variables.contains(&handle) {
+            unused.global_variables.insert(handle);
+        }
+    }
+    for (handle, _) in module.constants.iter() {
+        if !used_constants.contains(&handle) {
+            unused.constants.insert(handle);
+        }
+    }
+    for (handle, _) in module.functions.iter() {
+        if !used_functions.contains(&handle) {
+            unused.functions.insert(handle);
+        }
+    }
+
+    // Return the unused items
+    unused
+}
+
+/// Context for module crawling functions
+struct CrawlContext<'a> {
+    module: &'a crate::Module,
+    function: &'a crate::Function,
+    used_global_variables: &'a mut HashSet<Handle<crate::GlobalVariable>>,
+    used_types: &'a mut HashSet<Handle<crate::Type>>,
+    used_constants: &'a mut HashSet<Handle<crate::Constant>>,
+    used_functions: &'a mut HashSet<Handle<crate::Function>>,
+}
+
+/// Crawl a constant and register used items
+fn crawl_constant(constant: Handle<crate::Constant>, ctx: &mut CrawlContext) {
+    ctx.used_constants.insert(constant);
+    if let crate::ConstantInner::Composite { ty, ref components } =
+        ctx.module.constants[constant].inner
+    {
+        crawl_type(ty, &ctx.module.types, ctx.used_types);
+
+        for component in components {
+            crawl_constant(*component, ctx);
+        }
+    }
+}
+
+/// Crawl a type and register sub-types
+fn crawl_type(
+    ty: Handle<crate::Type>,
+    types: &crate::Arena<crate::Type>,
+    used_types: &mut HashSet<Handle<crate::Type>>,
+) {
+    used_types.insert(ty);
+
+    // Crawl any sub-types
+    match types[ty].inner {
+        crate::TypeInner::Scalar { kind: _, width: _ } => (),
```

Comment:
You can write `crate::TypeInner::Scalar { .. } => ...` if you not used any type fields.

Reply:
I was thinking maybe it would be good to do it like that so that if fields were added to the variant later it would cause an error. Do you think that's not really worth worrying about?

Reply:
I don't think we should worry about it here. But if we will, it would be great to add an extra comment with an explanation :)

Reply:
OK, I'll just switch to `..` then. :+1: It's probably over-paranoid to do it the other way. :)